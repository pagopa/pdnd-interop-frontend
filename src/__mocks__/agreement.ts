import { AgreementSummary } from '../../types'

export const agreementActiveEservicePublished: AgreementSummary = {
  id: 'dsfjds-jojoi-jdsfds',
  state: 'ACTIVE',
  eservice: {
    name: 'Riscossione TARI',
    id: 'osdijf-dsjfdisj-jsdfdsj',
    descriptorId: 'lskdfok-jisjdfs-djdsjfn',
    version: '4',
    state: 'PUBLISHED',
  },
  eserviceDescriptorId: 'dnsoifn-dsfjdsiof-dsjfsd',
  consumer: {
    name: 'Comune di Milano',
    id: 'afd_dss_cds',
  },
  producer: {
    name: 'Comune di Bologna',
    id: 'lds_kji_dsj',
  },
  attributes: [],
  suspendedByProducer: false,
  suspendedBySubscriber: false,
}

export const agreementSuspendedByProducerEservicePublished: AgreementSummary = {
  id: 'dsfjds-jojoi-jdsfds',
  state: 'SUSPENDED',
  eservice: {
    name: 'Riscossione TARI',
    id: 'osdijf-dsjfdisj-jsdfdsj',
    descriptorId: 'lskdfok-jisjdfs-djdsjfn',
    version: '4',
    state: 'PUBLISHED',
  },
  eserviceDescriptorId: 'dnsoifn-dsfjdsiof-dsjfsd',
  consumer: {
    name: 'Comune di Milano',
    id: 'afd_dss_cds',
  },
  producer: {
    name: 'Comune di Bologna',
    id: 'lds_kji_dsj',
  },
  attributes: [],
  suspendedByProducer: true,
  suspendedBySubscriber: false,
}

export const agreementSuspendedBySubscriberEservicePublished: AgreementSummary = {
  id: 'dsfjds-jojoi-jdsfds',
  state: 'SUSPENDED',
  eservice: {
    name: 'Riscossione TARI',
    id: 'osdijf-dsjfdisj-jsdfdsj',
    descriptorId: 'lskdfok-jisjdfs-djdsjfn',
    version: '4',
    state: 'PUBLISHED',
  },
  eserviceDescriptorId: 'dnsoifn-dsfjdsiof-dsjfsd',
  consumer: {
    name: 'Comune di Milano',
    id: 'afd_dss_cds',
  },
  producer: {
    name: 'Comune di Bologna',
    id: 'lds_kji_dsj',
  },
  attributes: [],
  suspendedByProducer: false,
  suspendedBySubscriber: true,
}
