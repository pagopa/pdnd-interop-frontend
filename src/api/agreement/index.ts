export * from './agreement.mutations'
export * from './agreement.queries'
export * from './agreement.downloads'
export * from './agreement.services'
