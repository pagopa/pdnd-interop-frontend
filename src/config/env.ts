import type { ExtendedWindow, PagoPAEnvVars } from '@/types/common.types'

const PAGOPA_ENV = (window as unknown as ExtendedWindow).pagopa_env

export const isDevelopment = !!(import.meta.env.MODE === 'development')
export const isProduction = !!(import.meta.env.MODE === 'production')
export const isTest = !!(import.meta.env.MODE === 'test')

if (!PAGOPA_ENV && !isTest) {
  console.warn('pagopa_env not available.')
}

const DEV_API_HOST_URL = import.meta.env.REACT_APP_API_HOST
const DEV_SELFCARE_LOGIN_URL = import.meta.env.REACT_APP_SELFCARE_LOGIN_URL
const DEV_INTEROP_RESOURCES_BASE_URL = import.meta.env.REACT_APP_INTEROP_RESOURCES_BASE_URL
const DEV_ONETRUST_DOMAIN_SCRIPT_ID = import.meta.env.REACT_APP_ONETRUST_DOMAIN_SCRIPT_ID

export const FE_LOGIN_URL = (
  isProduction && PAGOPA_ENV ? PAGOPA_ENV.SELFCARE_LOGIN_URL : DEV_SELFCARE_LOGIN_URL
) as string
export const PUBLIC_URL = import.meta.env.PUBLIC_URL || '/ui'
export const FE_URL = `${window.location.origin}${PUBLIC_URL}`

export const INTEROP_RESOURCES_BASE_URL =
  isProduction && PAGOPA_ENV
    ? PAGOPA_ENV.INTEROP_RESOURCES_BASE_URL
    : DEV_INTEROP_RESOURCES_BASE_URL
export const ONETRUST_DOMAIN_SCRIPT_ID = (
  isProduction && PAGOPA_ENV ? PAGOPA_ENV.ONETRUST_DOMAIN_SCRIPT_ID : DEV_ONETRUST_DOMAIN_SCRIPT_ID
) as string
export const MIXPANEL_PROJECT_ID = isProduction && PAGOPA_ENV ? PAGOPA_ENV.MIXPANEL_PROJECT_ID : ''

export const TEMP_USER_WHITELIST_URL = `${INTEROP_RESOURCES_BASE_URL}/temp-whitelist.json`

function getEnvVar(varName: keyof PagoPAEnvVars, devVarName: string): string {
  return isProduction && PAGOPA_ENV && PAGOPA_ENV
    ? PAGOPA_ENV[varName]
    : `${DEV_API_HOST_URL}/${devVarName}`
}

const SERVICE_VERSION = import.meta.env.REACT_APP_SERVICE_VERSION
export const API_GATEWAY_INTEFACE_URL = getEnvVar(
  'API_GATEWAY_INTEFACE_URL',
  'swagger/docs/interface-specification.yml'
)
export const BACKEND_FOR_FRONTEND_URL = getEnvVar(
  'BACKEND_FOR_FRONTEND_URL',
  `backend-for-frontend/${SERVICE_VERSION}`
)
export const AUTHORIZATION_PROCESS_URL = getEnvVar(
  'AUTHORIZATION_PROCESS_URL',
  `authorization-process/${SERVICE_VERSION}`
)
export const AUTHORIZATION_SERVER_ACCESS_TOKEN_URL = getEnvVar(
  'AUTHORIZATION_SERVER_TOKEN_CREATION_URL',
  'authorization-server/token.oauth2'
)
export const CLIENT_ASSERTION_JWT_AUDIENCE =
  isProduction && PAGOPA_ENV ? PAGOPA_ENV.CLIENT_ASSERTION_JWT_AUDIENCE : ''

function getWellKnownUrls(wellKnownUrls: string | undefined) {
  return wellKnownUrls?.split(',').filter((url) => !!url) || []
}

export const WELL_KNOWN_URLS =
  isProduction && PAGOPA_ENV ? getWellKnownUrls(PAGOPA_ENV.WELL_KNOWN_URLS) : ['#']

export const SELFCARE_BASE_URL =
  isProduction && PAGOPA_ENV ? PAGOPA_ENV.SELFCARE_BASE_URL : 'https://uat.selfcare.pagopa.it'

export const STAGE = PAGOPA_ENV?.STAGE ?? 'DEV'

export const SELFCARE_INTEROP_PROD_ID = `prod-interop${STAGE === 'TEST' ? '-coll' : ''}`
