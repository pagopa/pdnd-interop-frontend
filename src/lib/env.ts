import { ExtendedWindow, PagoPAEnvVars } from '../../types'

export const isDevelopment = !!(process.env.NODE_ENV === 'development')
export const isProduction = !!(process.env.NODE_ENV === 'production')

const PAGOPA_ENV = (window as unknown as ExtendedWindow).pagopa_env

const DEV_API_HOST_URL = process.env.REACT_APP_API_HOST
const DEV_SELFCARE_LOGIN_URL = process.env.REACT_APP_SELFCARE_LOGIN_URL
const DEV_INTEROP_RESOURCES_BASE_URL = process.env.REACT_APP_INTEROP_RESOURCES_BASE_URL
const DEV_ONETRUST_DOMAIN_SCRIPT_ID = process.env.REACT_APP_ONETRUST_DOMAIN_SCRIPT_ID

export const FE_LOGIN_URL = (
  isProduction ? PAGOPA_ENV.SELFCARE_LOGIN_URL : DEV_SELFCARE_LOGIN_URL
) as string
export const PUBLIC_URL = process.env.PUBLIC_URL
export const FE_URL = `${window.location.origin}${PUBLIC_URL}`

export const INTEROP_RESOURCES_BASE_URL = isProduction
  ? PAGOPA_ENV.INTEROP_RESOURCES_BASE_URL
  : DEV_INTEROP_RESOURCES_BASE_URL
export const ONETRUST_DOMAIN_SCRIPT_ID = (
  isProduction ? PAGOPA_ENV.ONETRUST_DOMAIN_SCRIPT_ID : DEV_ONETRUST_DOMAIN_SCRIPT_ID
) as string
export const MIXPANEL_PROJECT_ID = isProduction ? PAGOPA_ENV.MIXPANEL_PROJECT_ID : ''

export const TEMP_USER_WHITELIST_URL = `${INTEROP_RESOURCES_BASE_URL}/temp-whitelist.json`

function getEnvVar(varName: keyof PagoPAEnvVars, devVarName: string): string {
  return isProduction ? PAGOPA_ENV[varName] : `${DEV_API_HOST_URL}/${devVarName}`
}

const SERVICE_VERSION = process.env.REACT_APP_SERVICE_VERSION
export const API_GATEWAY_INTEFACE_URL = getEnvVar(
  'API_GATEWAY_INTEFACE_URL',
  'swagger/docs/interface-specification.yml'
)
export const BACKEND_FOR_FRONTEND_URL = getEnvVar(
  'BACKEND_FOR_FRONTEND_URL',
  `backend-for-frontend/${SERVICE_VERSION}`
)
export const AGREEMENT_PROCESS_URL = getEnvVar(
  'AGREEMENT_PROCESS_URL',
  `agreement-process/${SERVICE_VERSION}`
)
export const AUTHORIZATION_PROCESS_URL = getEnvVar(
  'AUTHORIZATION_PROCESS_URL',
  `authorization-process/${SERVICE_VERSION}`
)
export const CATALOG_PROCESS_URL = getEnvVar(
  'CATALOG_PROCESS_URL',
  `catalog-process/${SERVICE_VERSION}`
)
export const PURPOSE_PROCESS_URL = getEnvVar(
  'PURPOSE_PROCESS_URL',
  `purpose-process/${SERVICE_VERSION}`
)
export const AUTHORIZATION_SERVER_ACCESS_TOKEN_URL = getEnvVar(
  'AUTHORIZATION_SERVER_TOKEN_CREATION_URL',
  'authorization-server/token.oauth2'
)
export const CLIENT_ASSERTION_JWT_AUDIENCE = isProduction
  ? PAGOPA_ENV.CLIENT_ASSERTION_JWT_AUDIENCE
  : ''

function getWellKnownUrls(wellKnownUrls: string | undefined) {
  return wellKnownUrls?.split(',').filter((url) => !!url) || []
}

export const WELL_KNOWN_URLS = isProduction ? getWellKnownUrls(PAGOPA_ENV.WELL_KNOWN_URLS) : ['#']

// TEMP: Backend - This must come from the backend env
export const SELFCARE_BASE_URL = 'https://selfcare.pagopa.it'
