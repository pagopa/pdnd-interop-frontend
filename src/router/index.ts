export { RouterProvider } from './components/RouterProvider'
export { useCurrentRoute } from './hooks/useCurrentRoute'

import * as _routes from './routes'

export const { useNavigate, useParams, useLocation, useAuthGuard, useGeneratePath } = _routes.hooks
export const { Link, Redirect, Breadcrumbs } = _routes.components
export const { getParentRoutes } = _routes.utils
export type RouteKey = _routes.RouteKey
