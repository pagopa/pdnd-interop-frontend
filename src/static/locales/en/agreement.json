{
  "loadingSingleLabel": "Loading request",
  "loadingMultiLabel": "Loading requests",
  "backToRequestsBtn": "Go back to requests",
  "read": {
    "title": "Manage fruition request",
    "noPurposeAlert": "Your agreement request is active. <1>Create your first purpose</1>",
    "updateGuide": {
      "alertLabel": "A new version of <1>{{eserviceName}}</1> e-service is available. Before updating, read the guide below",
      "title": "Upgrade guide",
      "description": "It is possible to update the fruition request to the latest version of <1>{{eserviceName}}</1> e-service. Here are some useful informations",
      "upgradeGuideLinkLabel": "Upgrade fuide",
      "linksLabel": "Useful links",
      "eserviceLinkLabel": "E-service card <1>{{eserviceName}}, version {{version}}</1>",
      "upgradeBtn": "Upgrade"
    },
    "generalInformations": {
      "title": "General Informations",
      "eserviceField": {
        "label": "Request for",
        "goToEServiceBtn": "Go to e-service",
        "versionLabel": "version"
      },
      "providerField": {
        "label": "Provider",
        "infoLabel": "Max 1000 chars"
      },
      "consumerField": {
        "label": "Consumer"
      },
      "requestStatusField": {
        "label": "Request status"
      },
      "printableCopyField": {
        "label": "Printable copy",
        "docLabel": "Fruition request"
      }
    },
    "consumerMessage": {
      "title": "Message for provider"
    },
    "rejectionMessage": {
      "title": "Fruition request rejection reason",
      "description": "The provider states their reason for refusing to activate this fruition request. You can always submit a new one"
    },
    "attributes": {
      "howLink": "How?",
      "certified": {
        "title": "Certified attributes",
        "subtitle": "Questi attributi sono certificati da una fonte autoritativa riconosciuta, e non necessitano di ulteriori verifiche.",
        "emptyLabel": "There are no required certified attributes for this e-service"
      },
      "verified": {
        "title": "Verified attributes",
        "subtitle": "Questi attributi necessitano di verifica da parte di almeno un erogatore. Il fruitore potrebbe già possederne alcuni (o tutti).",
        "emptyLabel": "There are no required verified attributes for this e-service"
      },
      "declared": {
        "title": "Declared attributes",
        "subtitle": "Il fruitore deve dichiarare sotto la propria responsabilità di possedere questi attributi; potrebbe già possederne alcuni (o tutti).",
        "emptyLabel": "There are no required declared attributes for this e-service"
      }
    }
  },
  "edit": {
    "title": "Manage fruition request",
    "generalInformations": {
      "title": "General Informations",
      "eserviceField": {
        "label": "Request for",
        "goToEServiceBtn": "Go to e-service",
        "versionLabel": "version"
      },
      "providerField": {
        "label": "Provider"
      },
      "requestStatusField": {
        "label": "Request status"
      }
    },
    "attribute": {
      "subtitle": "You must own:",
      "certified": {
        "description": "These attributes are certified by a recognized authoritative source, and do not require further certification."
      },
      "verified": {
        "description": "These attributes require verification by at least one provider. You may already own some (or all) of them."
      },
      "declared": {
        "description": "You must declare under your own responsibility that you possess these attributes."
      }
    },
    "consumerNotes": {
      "title": "Message for provider",
      "description": "If there are any groups of attributes marked with “or”, indicate in this message which of those you want to be recognized, based on the documentation you have uploaded. Enter any other useful notes",
      "field": {
        "label": "Message",
        "infoLabel": "Max 1000 chars"
      }
    },
    "documents": {
      "title": "Documentation",
      "description": "Enter all the documents necessary for the recognition of your attributes",
      "uploadBtn": "Upload",
      "documentPrettynameField": {
        "label": "Document name",
        "infoLabel": "Min 5 chars, max 60 chars"
      }
    },
    "bottomPageActionCard": {
      "title": "Quick publish",
      "description": "Have you entered all the information for this agreement request? From here you can immediately publish a draft, or delete it. If you want to post later, just save the draft above or leave this page",
      "cancelBtn": "Delete draft",
      "submitBtn": "Publish agreement request",
      "updateBtn": "Save draft"
    },
    "backToRequestsBtn": "Go back to requests"
  },
  "list": {
    "eserviceName": "{{name}}, version {{version}}",
    "filters": {
      "eserviceField": {
        "label": "Find by e-service"
      },
      "statusField": {
        "label": "Agreement status",
        "optionLabels": {
          "ACTIVE": "Active",
          "DRAFT": "Draft",
          "SUSPENDED": "Suspended",
          "PENDING": "Pending",
          "ARCHIVED": "Archived",
          "REJECTED": "Rejected",
          "MISSING_CERTIFIED_ATTRIBUTES": "Missing certified attributes"
        }
      },
      "consumerField": {
        "label": "Find by consumer"
      },
      "providerField": {
        "label": "Find by provider"
      }
    },
    "provider": {
      "title": "Fruition requests",
      "description": "In this section, you can manage all the fruition requests you have received from parties that want to subscribe to your e-services"
    },
    "subscriber": {
      "title": "Your requests",
      "description": "In this section, you can manage all the fruition requests you have sent to e-service Providers"
    }
  }
}
