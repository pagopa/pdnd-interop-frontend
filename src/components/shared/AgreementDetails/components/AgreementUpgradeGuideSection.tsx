import { SectionContainer } from '@/components/layout/containers'
import { Alert, Box, Button, Divider, Link, Stack } from '@mui/material'
import React from 'react'
import { Trans, useTranslation } from 'react-i18next'
import LaunchIcon from '@mui/icons-material/Launch'
import LinkIcon from '@mui/icons-material/Link'
import { useAgreementDetailsContext } from '../AgreementDetailsContext'
import { AgreementMutations } from '@/api/agreement'
import { RouterLink, useNavigateRouter } from '@/router'
import { agreementUpgradeGuideLink } from '@/config/constants'
import { useJwt } from '@/hooks/useJwt'
import { InformationContainer } from '@pagopa/interop-fe-commons'

export const AgreementUpgradeGuideSection: React.FC = () => {
  const { t } = useTranslation('agreement', { keyPrefix: 'read.updateGuide' })
  const { agreement, canBeUpgraded } = useAgreementDetailsContext()
  const { mutate: upgradeAgreement } = AgreementMutations.useUpgrade()
  const { navigate } = useNavigateRouter()
  const { isAdmin } = useJwt()

  const handleUpgrade = async () => {
    if (!agreement?.id) return
    upgradeAgreement(
      { agreementId: agreement.id },
      {
        onSuccess() {
          navigate('SUBSCRIBE_AGREEMENT_LIST')
        },
      }
    )
  }

  const eservice = agreement?.eservice

  if (!eservice || !canBeUpgraded || !isAdmin) return null

  return (
    <>
      <Alert sx={{ mt: 2 }} severity="warning">
        <Trans components={{ 1: <Box component="span" fontWeight={700} /> }}>
          {t('alertLabel', { eserviceName: eservice.name })}
        </Trans>
      </Alert>

      <SectionContainer
        title={t('title')}
        description={
          <Trans components={{ 1: <Box component="span" fontWeight={700} /> }}>
            {t('description', { eserviceName: eservice.name })}
          </Trans>
        }
      >
        <Stack spacing={2}>
          <InformationContainer
            content={
              <Stack spacing={1}>
                <Link
                  component="a"
                  href={agreementUpgradeGuideLink}
                  target="_blank"
                  variant="body2"
                  underline="hover"
                  sx={{ display: 'flex', alignItems: 'center', cursor: 'pointer' }}
                >
                  <LaunchIcon fontSize="small" sx={{ mr: 1 }} /> {t('upgradeGuideLinkLabel')}
                </Link>
                <RouterLink
                  to="SUBSCRIBE_CATALOG_VIEW"
                  params={{ eserviceId: eservice.id, descriptorId: eservice.activeDescriptor!.id }}
                  variant="body2"
                  underline="hover"
                  target="_blank"
                  sx={{ display: 'flex', alignItems: 'center', cursor: 'pointer' }}
                >
                  <LinkIcon fontSize="small" sx={{ mr: 1 }} />{' '}
                  <span>
                    <Trans components={{ 1: <Box component="span" fontWeight={700} /> }}>
                      {t('eserviceLinkLabel', {
                        eserviceName: eservice.name,
                        version: eservice.activeDescriptor?.version,
                      })}
                    </Trans>
                  </span>
                </RouterLink>
              </Stack>
            }
            label={t('linksLabel')}
          />
          <Divider />
          <Stack direction="row" justifyContent="center">
            <Button onClick={handleUpgrade} variant="outlined">
              {t('upgradeBtn')}
            </Button>
          </Stack>
        </Stack>
      </SectionContainer>
    </>
  )
}
