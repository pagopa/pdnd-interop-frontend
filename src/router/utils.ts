import { routes } from '@/router/routes'
import type { LangCode, ProviderOrConsumer } from '@/types/common.types'
import type { RouteConfig, RouteKey } from '@/router/types'
import { generatePath, matchPath } from 'react-router-dom'
import { getKeys } from '@/utils/array.utils'
import memoize from 'lodash/memoize'
import identity from 'lodash/identity'
import sortBy from 'lodash/sortBy'
import { PUBLIC_URL } from '@/config/env'
import qs from 'qs'
import QueryString from 'qs'

/** Returns the localized path of the given routeKey and language  */
export function getLocalizedRoutePathname(routeKey: RouteKey, lang: LangCode) {
  return `${PUBLIC_URL}/${lang}/${routes[routeKey].PATH[lang]}`
}

/** Checks if the routeKey represent the given pathname */
function matchRouteKeyPath(pathname: string, lang: LangCode, routeKey: RouteKey) {
  return matchPath(getLocalizedRoutePathname(routeKey as RouteKey, lang), pathname)
}

/** Returns the routeKey of the given pathname */
export const getRouteKeyFromPathname = memoize(function _getRouteKeyFromPathname(
  pathname: string,
  lang: LangCode
) {
  let pathnameWithBaseUrl = pathname
  if (!pathnameWithBaseUrl.startsWith(PUBLIC_URL)) {
    pathnameWithBaseUrl = PUBLIC_URL + pathnameWithBaseUrl
  }
  const isTheRouteKeyPath = matchRouteKeyPath.bind(null, pathnameWithBaseUrl, lang)
  const routeKey = getKeys(routes).find(isTheRouteKeyPath)

  if (!routeKey) {
    throw new Error(`Pathname ${pathnameWithBaseUrl} has no associated routeKey.`)
  }

  return routeKey
})

export function switchPathLang(fromLang: LangCode, toLang: LangCode) {
  for (const routeKey of Object.keys(routes)) {
    const match = matchRouteKeyPath(window.location.pathname, fromLang, routeKey as RouteKey)

    if (match) {
      let newPath = generatePath(
        getLocalizedRoutePathname(routeKey as RouteKey, toLang),
        match.params as Record<string, string>
      )

      if (window.location.search) {
        newPath += window.location.search
      }

      if (window.location.hash) {
        newPath += window.location.hash
      }

      window.location.replace(newPath)
      return
    }
  }
}

export const URL_FRAGMENTS: Record<string, Record<LangCode, string>> = {
  FIRST_DRAFT: { it: 'prima-bozza', en: 'first-draft' },
  EDIT: { it: 'modifica', en: 'bozza' },
}

export function getSplittedPath(route: RouteConfig, currentLang: LangCode) {
  return route.PATH[currentLang].split('/').filter(identity)
}

export const getParentRoutes = memoize((routeKey: RouteKey): Array<RouteKey> => {
  function isParentRoute(
    possibleParentRouteSubpaths: Array<string>,
    currentSubpaths: Array<string>
  ) {
    if (possibleParentRouteSubpaths.length >= currentSubpaths.length) {
      return false
    }

    const allSameFragments = possibleParentRouteSubpaths.every(
      (pathFragment, i) => pathFragment === currentSubpaths[i]
    )

    // URL_FRAGMENTS.EDIT is appended at the end of a read path of the same type.
    // E.g. /eservice/myid becomes /eservice/myid/edit. So it's always same length + 1
    const isFalseParent = currentSubpaths.length === possibleParentRouteSubpaths.length + 1

    if (allSameFragments && isEditPath(routeKey) && isFalseParent) {
      return false
    }

    return allSameFragments
  }

  const route = routes[routeKey]
  const currentSubpaths = getSplittedPath(route, 'it')

  const parents = Object.entries(routes).filter(([_, possibleParentRoute]) =>
    isParentRoute(getSplittedPath(possibleParentRoute, 'it'), currentSubpaths)
  )

  const sortedParents = sortBy(parents, ([_, parent]) => getSplittedPath(parent, 'it').length)

  return sortedParents.map(([routeKey]) => routeKey) as Array<RouteKey>
})

export const isProviderOrConsumerRoute = memoize(
  (routeKey: RouteKey): ProviderOrConsumer | null => {
    const excludeList = ['ui', 'it']
    const subroutes = getSplittedPath(routes[routeKey], 'it').filter(
      (b) => !excludeList.includes(b)
    )
    const [mode] = subroutes

    if (mode === 'erogazione') {
      return 'provider'
    }

    if (mode === 'fruizione') {
      return 'consumer'
    }

    return null
  }
)

export const isEditPath = memoize((routeKey: RouteKey): boolean => {
  const subroutes = getSplittedPath(routes[routeKey], 'it')

  const lastBit = subroutes[subroutes.length - 1]
  return Object.values(URL_FRAGMENTS.EDIT).some((f) => lastBit.endsWith(f))
})

export function parseSearch(search: string) {
  return qs.parse(search, { ignoreQueryPrefix: true })
}

export function stringifySearch(searchObj: Record<string, string> | QueryString.ParsedQs) {
  return qs.stringify(searchObj)
}

/**
 * Returns an array with all the dynamic path names for a given RouteKey
 * @example
 * "/:foo/test/:bar" => ["foo", "bar"]
 */
export const getDynamicPathsName = memoize((routeKey: RouteKey) => {
  return routes[routeKey].PATH.it
    .split('/')
    .filter(identity)
    .filter((subpath) => subpath.startsWith(':'))
    .map((param) => param.replace(':', ''))
})
