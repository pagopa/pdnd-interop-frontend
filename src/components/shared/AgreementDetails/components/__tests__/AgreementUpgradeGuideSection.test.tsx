import React from 'react'
import { AgreementUpgradeGuideSection } from '../AgreementUpgradeGuideSection'
import { createMockAgreement } from '__mocks__/data/agreement.mocks'
import { mockUseJwt, renderWithApplicationContext } from '@/utils/testing.utils'
import { mockAgreementDetailsContext } from './test.commons'
import { vi } from 'vitest'
import { setupServer } from 'msw/node'
import { rest } from 'msw'
import { BACKEND_FOR_FRONTEND_URL } from '@/config/env'
import userEvent from '@testing-library/user-event'
import { waitFor } from '@testing-library/react'
import * as router from '@/router'

const server = setupServer(
  rest.post(`${BACKEND_FOR_FRONTEND_URL}/agreements/:agreementId/upgrade`, (req, res, ctx) => {
    return res(ctx.status(200))
  })
)

beforeAll(() => server.listen())
afterAll(() => server.close())

const useNavigateSpy = vi.spyOn(router, 'useNavigateRouter')

describe('AgreementUpgradeGuideSection', () => {
  it('should match the snapshot', () => {
    mockUseJwt({ isAdmin: true })
    mockAgreementDetailsContext({
      agreement: createMockAgreement(),
      canBeUpgraded: true,
    })

    const { baseElement } = renderWithApplicationContext(<AgreementUpgradeGuideSection />, {
      withReactQueryContext: true,
      withRouterContext: true,
    })

    expect(baseElement).toMatchSnapshot()
  })

  it('should not render if user is not an admin', () => {
    mockUseJwt({ isAdmin: false })
    mockAgreementDetailsContext({
      agreement: createMockAgreement(),
      canBeUpgraded: true,
    })

    const { container } = renderWithApplicationContext(<AgreementUpgradeGuideSection />, {
      withReactQueryContext: true,
      withRouterContext: true,
    })

    expect(container).toBeEmptyDOMElement()
  })

  it('should not render if the agreement cannot be upgraded', () => {
    mockUseJwt({ isAdmin: true })
    mockAgreementDetailsContext({
      agreement: createMockAgreement(),
      canBeUpgraded: false,
    })

    const { container } = renderWithApplicationContext(<AgreementUpgradeGuideSection />, {
      withReactQueryContext: true,
      withRouterContext: true,
    })

    expect(container).toBeEmptyDOMElement()
  })

  it('should redirect to the agreement list after upgrading', async () => {
    const navigateRouterFn = vi.fn()
    useNavigateSpy.mockReturnValue({ navigate: navigateRouterFn, getRouteUrl: () => '' })

    mockUseJwt({ isAdmin: true })
    mockAgreementDetailsContext({
      agreement: createMockAgreement(),
      canBeUpgraded: true,
    })

    const screen = renderWithApplicationContext(<AgreementUpgradeGuideSection />, {
      withReactQueryContext: true,
      withRouterContext: true,
    })

    const user = userEvent.setup()
    await user.click(screen.getByRole('button', { name: 'upgradeBtn' }))
    await user.click(screen.getByRole('button', { name: /confirm/i }))

    await waitFor(() => {
      expect(navigateRouterFn).toBeCalledWith('SUBSCRIBE_AGREEMENT_LIST')
    })
  })
})
