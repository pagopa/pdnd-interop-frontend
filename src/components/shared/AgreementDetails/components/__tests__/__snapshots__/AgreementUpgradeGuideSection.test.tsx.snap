// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AgreementUpgradeGuideSection > should match the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardWarning MuiAlert-standard css-1ot7o2q-MuiPaper-root-MuiAlert-root"
      role="alert"
    >
      <div
        class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="ReportProblemOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
          />
        </svg>
      </div>
      <div
        class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
      >
        alertLabel
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1w5x330-MuiPaper-root"
    >
      <div
        class="css-nen11g-MuiStack-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
        >
          title
        </h2>
        <span
          class="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
        >
          description
        </span>
      </div>
      <div
        class="MuiBox-root css-1yuhvjn"
      >
        <div
          class="css-1p5q5e5-MuiStack-root"
        >
          <div
            class="css-19qjufr-MuiStack-root"
          >
            <div
              class="MuiBox-root css-iuqs4u"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
              >
                linksLabel
              </p>
            </div>
            <div
              class="MuiBox-root css-1rr4qq7"
            >
              <div
                class="MuiTypography-root MuiTypography-body2 css-hoiwr4-MuiTypography-root"
              >
                <div
                  class="css-1okj3ks-MuiStack-root"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-1rq2pt3-MuiTypography-root-MuiLink-root"
                    href="https://docs.pagopa.it/interoperabilita-1/manuale-operativo/richieste-di-fruizione#fruitore-aggiornare-una-richiesta-di-fruizione"
                    target="_blank"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-15ikgma-MuiSvgIcon-root"
                      data-testid="LaunchIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                      />
                    </svg>
                     
                    upgradeGuideLinkLabel
                  </a>
                  <a
                    class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineHover css-1rq2pt3-MuiTypography-root-MuiLink-root"
                    href="/it/fruizione/catalogo-e-service/47f82055-77fd-4efd-8ed6-8a7d3021c879/b79fc9ac-2882-49bd-afd1-71f4284f117c"
                    target="_blank"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-15ikgma-MuiSvgIcon-root"
                      data-testid="LinkIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
                      />
                    </svg>
                     
                    <span>
                      eserviceLinkLabel
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="css-13f3otb-MuiStack-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              upgradeBtn
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
